# Assignment 5

For BABI 4005 DQM with Python

By: ITMG 4J Sijian Liang (Cindy)

_Updated 2025-02-18_

## Summary

In this assignment, I found another dataset that can join my original data and finished merging to answer some questions.

The original dataset is the Heart Attack Risk Dataset: https://www.kaggle.com/datasets/arifmia/heart-attack-risk-dataset

The new dataset is Obesity Prediction Dataset: https://www.kaggle.com/datasets/ruchikakumbhar/obesity-prediction

## Data Merging Process

1. Import libraries and datasets with updating paths since the file structure on Github has been changed.
2. Observing data summaries to ensure merge feasibility
3. Find common keys between both datasets and explain the reasons
4. Convert data types to ensure merging without errors
5. Remove duplicates BEFORE merging
6. Merge datasets with common keys and explain the merging way and reasons


### 1. Import Libraries and Datasets

First, import those libraries and datasets.

In [4]:
import pandas as pd
import numpy as np

# Since the file structure has been updated, I updated the paths.
# Load the original heart attack dataset
heart_attack= pd.read_csv("../data/heart_attack_risk_dataset.csv")

# Load the new obesity dataset
obesity= pd.read_csv("../data/Obesity prediction.csv")


#### Check If I can access the dataset after updating the path

1) After getting the first 3 rows from the Heart attack dataset, I found that the path was correct and I could access the dataset.

In [6]:
## Get the first 3 rows from the dataset
heart_attack.head(3)

Age  Gender  Smoking  Alcohol_Consumption Physical_Activity_Level    BMI  \
0   69  Female        1                    0                Moderate  34.61   
1   32    Male        0                    0                Moderate  22.75   
2   89    Male        0                    1                Moderate  35.32   

   Diabetes  Hypertension  Cholesterol_Level  Resting_BP  Heart_Rate  \
0         1             0              152.1         171          85   
1         0             0              166.8         126         103   
2         0             0              272.3         123         127   

   Family_History Stress_Level Chest_Pain_Type        Thalassemia  \
0               0     Moderate     Non-anginal  Reversible defect   
1               0          Low    Asymptomatic             Normal   
2               0          Low         Typical  Reversible defect   

   Fasting_Blood_Sugar       ECG_Results  Exercise_Induced_Angina  \
0                    0            Normal                        0   
1                    0  ST-T abnormality                        0   
2                    0  ST-T abnormality                        0   

   Max_Heart_Rate_Achieved Heart_Attack_Risk  
0                      114               Low  
1                      173          Moderate  
2                      109               Low

2) Did the same action to getting the first 3 rows from the obesity dataset

In [8]:
## Get the first 10 rows from the dataset
obesity.head(3)

Gender   Age  Height  Weight Family_History FAVC  FCVC  NCP       CAEC  \
0  Female  21.0    1.62    64.0            yes   no   2.0  3.0  Sometimes   
1  Female  21.0    1.52    56.0            yes   no   3.0  3.0  Sometimes   
2    Male  23.0    1.80    77.0            yes   no   2.0  3.0  Sometimes   

  Smoking  CH2O  SCC  FAF  TUE        CALC                 MTRANS  \
0      no   2.0   no  0.0  1.0          no  Public_Transportation   
1     yes   3.0  yes  3.0  0.0   Sometimes  Public_Transportation   
2      no   2.0   no  2.0  1.0  Frequently  Public_Transportation   

         Obesity  
0  Normal_Weight  
1  Normal_Weight  
2  Normal_Weight

## 2. Observe the Dataset Summaries to Ensure Merge Feasibility

- The heart attack dataset has 50000 rows.

- The Obesity dataset has 2111 rows.

- Both of datasets have zero "Nan" values.

- Both of them have "Age" and "Gender" Columns.

- Both of them's "Gender" columns only have two unique values.

- Therefore, doing data merging seems feasible.

In [10]:
# Data summary of Heart Attack Dataset
heart_attack.describe(include = "all")

Age  Gender       Smoking  Alcohol_Consumption  \
count   50000.000000   50000  50000.000000         50000.000000   
unique           NaN       2           NaN                  NaN   
top              NaN  Female           NaN                  NaN   
freq             NaN   25086           NaN                  NaN   
mean       53.398700     NaN      0.298420             0.401980   
std        20.799006     NaN      0.457569             0.490303   
min        18.000000     NaN      0.000000             0.000000   
25%        35.000000     NaN      0.000000             0.000000   
50%        53.000000     NaN      0.000000             0.000000   
75%        71.000000     NaN      1.000000             1.000000   
max        89.000000     NaN      1.000000             1.000000   

       Physical_Activity_Level           BMI      Diabetes  Hypertension  \
count                    50000  50000.000000  50000.000000   50000.00000   
unique                       3           NaN           NaN           NaN   
top                        Low           NaN           NaN           NaN   
freq                     20061           NaN           NaN           NaN   
mean                       NaN     27.519357      0.198840       0.29738   
std                        NaN      7.225176      0.399131       0.45711   
min                        NaN     15.000000      0.000000       0.00000   
25%                        NaN     21.270000      0.000000       0.00000   
50%                        NaN     27.540000      0.000000       0.00000   
75%                        NaN     33.770000      0.000000       1.00000   
max                        NaN     40.000000      1.000000       1.00000   

        Cholesterol_Level    Resting_BP    Heart_Rate  Family_History  \
count        50000.000000  50000.000000  50000.000000     50000.00000   
unique                NaN           NaN           NaN             NaN   
top                   NaN           NaN           NaN             NaN   
freq                  NaN           NaN           NaN             NaN   
mean           225.036444    134.434100     94.530480         0.30034   
std             43.317373     25.907961     20.187124         0.45841   
min            150.000000     90.000000     60.000000         0.00000   
25%            187.300000    112.000000     77.000000         0.00000   
50%            225.100000    134.000000     94.000000         0.00000   
75%            262.600000    157.000000    112.000000         1.00000   
max            300.000000    179.000000    129.000000         1.00000   

       Stress_Level Chest_Pain_Type Thalassemia  Fasting_Blood_Sugar  \
count         50000           50000       50000         50000.000000   
unique            3               4           3                  NaN   
top        Moderate         Typical      Normal                  NaN   
freq          24971           12642       16797                  NaN   
mean            NaN             NaN         NaN             0.149360   
std             NaN             NaN         NaN             0.356447   
min             NaN             NaN         NaN             0.000000   
25%             NaN             NaN         NaN             0.000000   
50%             NaN             NaN         NaN             0.000000   
75%             NaN             NaN         NaN             0.000000   
max             NaN             NaN         NaN             1.000000   

       ECG_Results  Exercise_Induced_Angina  Max_Heart_Rate_Achieved  \
count        50000             50000.000000             50000.000000   
unique           3                      NaN                      NaN   
top         Normal                      NaN                      NaN   
freq         16799                      NaN                      NaN   
mean           NaN                 0.201640               149.305340   
std            NaN                 0.401229                28.823699   
min            NaN                 0.000000          

In [11]:
heart_attack.isna().sum()

Age                        0
Gender                     0
Smoking                    0
Alcohol_Consumption        0
Physical_Activity_Level    0
BMI                        0
Diabetes                   0
Hypertension               0
Cholesterol_Level          0
Resting_BP                 0
Heart_Rate                 0
Family_History             0
Stress_Level               0
Chest_Pain_Type            0
Thalassemia                0
Fasting_Blood_Sugar        0
ECG_Results                0
Exercise_Induced_Angina    0
Max_Heart_Rate_Achieved    0
Heart_Attack_Risk          0
dtype: int64

In [12]:
# Data summary of Obesity Dataset
obesity.describe(include = "all")

Gender          Age       Height       Weight Family_History  FAVC  \
count    2111  2111.000000  2111.000000  2111.000000           2111  2111   
unique      2          NaN          NaN          NaN              2     2   
top      Male          NaN          NaN          NaN            yes   yes   
freq     1068          NaN          NaN          NaN           1726  1866   
mean      NaN    24.312600     1.701677    86.586058            NaN   NaN   
std       NaN     6.345968     0.093305    26.191172            NaN   NaN   
min       NaN    14.000000     1.450000    39.000000            NaN   NaN   
25%       NaN    19.947192     1.630000    65.473343            NaN   NaN   
50%       NaN    22.777890     1.700499    83.000000            NaN   NaN   
75%       NaN    26.000000     1.768464   107.430682            NaN   NaN   
max       NaN    61.000000     1.980000   173.000000            NaN   NaN   

               FCVC          NCP       CAEC Smoking         CH2O   SCC  \
count   2111.000000  2111.000000       2111    2111  2111.000000  2111   
unique          NaN          NaN          4       2          NaN     2   
top             NaN          NaN  Sometimes      no          NaN    no   
freq            NaN          NaN       1765    2067          NaN  2015   
mean       2.419043     2.685628        NaN     NaN     2.008011   NaN   
std        0.533927     0.778039        NaN     NaN     0.612953   NaN   
min        1.000000     1.000000        NaN     NaN     1.000000   NaN   
25%        2.000000     2.658738        NaN     NaN     1.584812   NaN   
50%        2.385502     3.000000        NaN     NaN     2.000000   NaN   
75%        3.000000     3.000000        NaN     NaN     2.477420   NaN   
max        3.000000     4.000000        NaN     NaN     3.000000   NaN   

                FAF          TUE       CALC                 MTRANS  \
count   2111.000000  2111.000000       2111                   2111   
unique          NaN          NaN          4                      5   
top             NaN          NaN  Sometimes  Public_Transportation   
freq            NaN          NaN       1401                   1580   
mean       1.010298     0.657866        NaN                    NaN   
std        0.850592     0.608927        NaN                    NaN   
min        0.000000     0.000000        NaN                    NaN   
25%        0.124505     0.000000        NaN                    NaN   
50%        1.000000     0.625350        NaN                    NaN   
75%        1.666678     1.000000        NaN                    NaN   
max        3.000000     2.000000        NaN                    NaN   

               Obesity  
count             2111  
unique               7  
top     Obesity_Type_I  
freq               351  
mean               NaN  
std                NaN  
min                NaN  
25%                NaN  
50%                NaN  
75%                NaN  
max                NaN

In [13]:
obesity.isna().sum()

Gender            0
Age               0
Height            0
Weight            0
Family_History    0
FAVC              0
FCVC              0
NCP               0
CAEC              0
Smoking           0
CH2O              0
SCC               0
FAF               0
TUE               0
CALC              0
MTRANS            0
Obesity           0
dtype: int64

## 3. Found the common keys between these two datasets

In these two datasets, "Age" and "Gender" can become the common keys for data merging.

### - Why Use Age and Gender to Join?

*Reason 1*: Both datasets contain Age and Gender. This makes them natural candidates for merging.

*Reason 2*: Medical and health data often rely on Age and Gender.

- Age is a major factor in heart disease and obesity.
- Gender plays a role in risk factors (e.g., men generally have a higher risk of heart attacks at a younger age, while women may have different obesity patterns).
- Merging on these ensures that I am matching similar individuals in both datasets.

### - Before merging, datatypes should be confirmed.

However, before merging, I need to ensure that the common keys are the same datatypes in both datasets in case there will be merging errors. Otherwise, I got an error saying "*py:2: UserWarning: You are merging on int and float columns where the float values are not equal to their int representation.*"


## 4. Convert data types to ensure merging without errors

### 4.1  Check the column data types of the Heart Attack Dataset

Based on the findings in Assignment 3 - baby project, this dataset was clean. 

The only issue was "Smoking", "Alcohol_Consumption", "Diabetes", "Hypertension", "Hypertension", "Family_History", "Fasting_Blood_Sugar", "Exercise_Induced_Angina" columns should be boolean data types.

To ensure the further merging without problem, I converted these 7 columns to boolean data types. 

In [17]:
## Check the column data types to ensure merging datasets successfully 
heart_attack.dtypes

Age                          int64
Gender                      object
Smoking                      int64
Alcohol_Consumption          int64
Physical_Activity_Level     object
BMI                        float64
Diabetes                     int64
Hypertension                 int64
Cholesterol_Level          float64
Resting_BP                   int64
Heart_Rate                   int64
Family_History               int64
Stress_Level                object
Chest_Pain_Type             object
Thalassemia                 object
Fasting_Blood_Sugar          int64
ECG_Results                 object
Exercise_Induced_Angina      int64
Max_Heart_Rate_Achieved      int64
Heart_Attack_Risk           object
dtype: object

### 4.2 Convert those Seven Columns of the Heart Attack Dataset to Boolean

Seven columns, including "Smoking", "Alcohol_Consumption", "Diabetes", "Hypertension","Family_History", "Fasting_Blood_Sugar", "Exercise_Induced_Angina", should only show "True" or "False".


In [19]:
# Convert 7 int64 columns to boolean variables
# Including Smoking, Alcohol_Consumption, Diabetes, Hypertension, Hypertension, Family_History, Fasting_Blood_Sugar, Exercise_Induced_Angina

heart_attack['Smoking'] = heart_attack['Smoking'].apply(lambda x: True if x == 1 else False)
heart_attack['Alcohol_Consumption'] = heart_attack['Alcohol_Consumption'].apply(lambda x: True if x == 1 else False)
heart_attack['Diabetes'] = heart_attack['Diabetes'].apply(lambda x: True if x == 1 else False)
heart_attack['Hypertension'] = heart_attack['Hypertension'].apply(lambda x: True if x == 1 else False)
heart_attack['Family_History'] = heart_attack['Family_History'].apply(lambda x: True if x == 1 else False)
heart_attack['Fasting_Blood_Sugar'] = heart_attack['Fasting_Blood_Sugar'].apply(lambda x: True if x == 1 else False)
heart_attack['Exercise_Induced_Angina'] = heart_attack['Exercise_Induced_Angina'].apply(lambda x: True if x == 1 else False)

heart_attack.head(10)

Age  Gender  Smoking  Alcohol_Consumption Physical_Activity_Level    BMI  \
0   69  Female     True                False                Moderate  34.61   
1   32    Male    False                False                Moderate  22.75   
2   89    Male    False                 True                Moderate  35.32   
3   78    Male    False                 True                Moderate  18.23   
4   38  Female     True                False                Moderate  19.82   
5   41    Male    False                 True                Moderate  36.11   
6   20    Male     True                False                     Low  15.12   
7   39  Female    False                False                Moderate  20.78   
8   70    Male    False                 True                    High  29.20   
9   19  Female    False                False                    High  35.54   

   Diabetes  Hypertension  Cholesterol_Level  Resting_BP  Heart_Rate  \
0      True         False              152.1         171          85   
1     False         False              166.8         126         103   
2     False         False              272.3         123         127   
3      True         False              237.7         144         125   
4     False         False              207.7         123         107   
5     False         False              271.2         141         119   
6     False         False              164.8         154          67   
7     False         False              297.0          91         112   
8     False         False              280.7         121          71   
9     False          True              275.0         167          76   

   Family_History Stress_Level Chest_Pain_Type        Thalassemia  \
0           False     Moderate     Non-anginal  Reversible defect   
1           False          Low    Asymptomatic             Normal   
2           False          Low         Typical  Reversible defect   
3           False          Low         Typical       Fixed defect   
4           False         High    Asymptomatic  Reversible defect   
5           False          Low        Atypical       Fixed defect   
6           False          Low     Non-anginal             Normal   
7           False          Low        Atypical             Normal   
8           False     Moderate     Non-anginal       Fixed defect   
9           False          Low         Typical  Reversible defect   

   Fasting_Blood_Sugar                   ECG_Results  Exercise_Induced_Angina  \
0                False                        Normal                    False   
1                False              ST-T abnormality                    False   
2                False              ST-T abnormality                    False   
3                 True  Left Ventricular Hypertrophy                    False   
4                False              ST-T abnormality                    False   
5                False                        Normal                    False   
6                False                        Normal                    False   
7                False              ST-T abnormality                     True   
8                False              ST-T abnormality                    False   
9                 True  Left Ventricular Hypertrophy                    False   

   Max_Heart_Rate_Achieved Heart_Attack_Risk  
0                      114               Low  
1                      173          Moderate  
2                      109               Low  
3                      129               Low  
4                      124          Moderate  
5                      101              High  
6                      176               Low  
7                      134          Moderate  
8                      168          Moderate  
9                      102          Moderate

### 4.3 Check the column data types of the Obesity Dataset

After checking the data types of the dataset, I found that the "Family_History", "FAVC", "Smoking" and "SCC" four columns should be boolean variables.

In addition, "Age" is "float64". However, in the heart attack dataset "Age" is "int64".

Therefore, I converted those four columns to boolean and "Age" to "int64" to ensure further merging without problems.

In [21]:
## Check the column data types to ensure merging datasets successfully 
obesity.dtypes

Gender             object
Age               float64
Height            float64
Weight            float64
Family_History     object
FAVC               object
FCVC              float64
NCP               float64
CAEC               object
Smoking            object
CH2O              float64
SCC                object
FAF               float64
TUE               float64
CALC               object
MTRANS             object
Obesity            object
dtype: object

### 4.4 Convert Obesity Dataset's four Columns to Boolean Variables

Try to use the function I wrote in assignment 4 to clean the obesity dataset and convert "Family_History", "FAVC", "Smoking" and "SCC" to boolean values.

In [23]:
def clean_column_types(df, num_cols, str_cols, bool_cols):
    """
    Ensure columns have the correct data types:
    - Convert numerical columns to numeric (coerce invalid values to NaN)
    - Convert string columns to string type
    - Convert boolean columns to boolean type
    """
     # Convert numerical columns
    df[num_cols] = df[num_cols].apply(pd.to_numeric, errors='coerce')  # Force numeric conversion

    # Convert string columns
    df[str_cols] = df[str_cols].astype(str).apply(lambda x: x.str.strip())  # Strip spaces
    
   
    # Define the boolean conversion function
    def convert_to_bool(value):
        if pd.isna(value):  # Handle NaN cases
            return np.nan
        value = str(value).strip().lower()  # Normalize case and remove spaces
        if value in ['yes', '1', 'true', 'y', 't']:  # Accept different True formats
            return True
        elif value in ['no', '0', 'false', 'n', 'f']:  # Accept different False formats
            return False
        else:
            return np.nan  # If value is unknown, return NaN

    # Apply conversion function to boolean columns using apply() instead of map()
    for col in bool_cols:
        df[col] = df[col].astype(str).str.strip().apply(convert_to_bool)

    return df


In [24]:
# Define column categories
num_cols = ["Age", "Height", "Weight", "FCVC", "NCP", "CH2O", 
            "FAF", "TUE"]
str_cols = ["Gender", "CAEC", "CALC","MTRANS","Obesity"]
bool_cols = ["Family_History", "FAVC", "Smoking","SCC"]

# Apply the function to clean column types
obesity_clean = clean_column_types(obesity, num_cols, str_cols, bool_cols)

#### After cleaning, check the Obesity Dataset's data types and first 10 rows again

To confirm if the data have been adjusted

In [26]:
# After cleaning, check the data types again
obesity_clean.dtypes

Gender             object
Age               float64
Height            float64
Weight            float64
Family_History       bool
FAVC                 bool
FCVC              float64
NCP               float64
CAEC               object
Smoking              bool
CH2O              float64
SCC                  bool
FAF               float64
TUE               float64
CALC               object
MTRANS             object
Obesity            object
dtype: object

### 4.5 Covert the "Age" columns to same data types

After checking the two datasets, I found that the "Age" in the Heart Attack dataset is "Int64" while the "Age" in the Obesity dataset is "float64".

Therefore, I used pandas.DataFrame.astype() function to adjust column types, changing the "Age" data type in the Obesity dataset to "Int64".

In [28]:
# Use the pandas.DataFrame.astype(<type>) function to manipulate column dtypes.
# https://stackoverflow.com/questions/21291259/convert-floats-to-ints-in-pandas

obesity['Age'] = obesity['Age'].astype(int)

In [29]:
# Recheck the data types after converting
obesity.dtypes 

Gender             object
Age                 int32
Height            float64
Weight            float64
Family_History       bool
FAVC                 bool
FCVC              float64
NCP               float64
CAEC               object
Smoking              bool
CH2O              float64
SCC                  bool
FAF               float64
TUE               float64
CALC               object
MTRANS             object
Obesity            object
dtype: object

In [30]:
# Check the first 10 rows of clean obesity data 
obesity_clean.head(10)

Gender  Age  Height  Weight  Family_History   FAVC  FCVC  NCP       CAEC  \
0  Female   21    1.62    64.0            True  False   2.0  3.0  Sometimes   
1  Female   21    1.52    56.0            True  False   3.0  3.0  Sometimes   
2    Male   23    1.80    77.0            True  False   2.0  3.0  Sometimes   
3    Male   27    1.80    87.0           False  False   3.0  3.0  Sometimes   
4    Male   22    1.78    89.8           False  False   2.0  1.0  Sometimes   
5    Male   29    1.62    53.0           False   True   2.0  3.0  Sometimes   
6  Female   23    1.50    55.0            True   True   3.0  3.0  Sometimes   
7    Male   22    1.64    53.0           False  False   2.0  3.0  Sometimes   
8    Male   24    1.78    64.0            True   True   3.0  3.0  Sometimes   
9    Male   22    1.72    68.0            True   True   2.0  3.0  Sometimes   

   Smoking  CH2O    SCC  FAF  TUE        CALC                 MTRANS  \
0    False   2.0  False  0.0  1.0          no  Public_Transportation   
1     True   3.0   True  3.0  0.0   Sometimes  Public_Transportation   
2    False   2.0  False  2.0  1.0  Frequently  Public_Transportation   
3    False   2.0  False  2.0  0.0  Frequently                Walking   
4    False   2.0  False  0.0  0.0   Sometimes  Public_Transportation   
5    False   2.0  False  0.0  0.0   Sometimes             Automobile   
6    False   2.0  False  1.0  0.0   Sometimes              Motorbike   
7    False   2.0  False  3.0  0.0   Sometimes  Public_Transportation   
8    False   2.0  False  1.0  1.0  Frequently  Public_Transportation   
9    False   2.0  False  1.0  1.0          no  Public_Transportation   

               Obesity  
0        Normal_Weight  
1        Normal_Weight  
2        Normal_Weight  
3   Overweight_Level_I  
4  Overweight_Level_II  
5        Normal_Weight  
6        Normal_Weight  
7        Normal_Weight  
8        Normal_Weight  
9        Normal_Weight

## 5. Remove Duplicates BEFORE Merging

In step 1, I found that both of these two datasets have duplicate records per Age-Gender.

This would distort the analysis and misrepresent the relationship between heart attack risk and obesity level because duplicated records will inflate correlations and statistical significance incorrectly.

Therefore, removing duplicates before merging is important because it prevents the many-to-many explosion. 

But I didn't drop duplicates of any dataset because it may result in data loss and data integrity issues (If multiple people in the same Age-Gender group have different obesity levels, removing duplicates may lead to incorrect conclusions.)

To analyze the relationship between heart attack risk and obesity, the best way to do it is: "Aggregate each dataset to summarize key metrics so that each Age-Gender pair has one unique record. " 

Thus, I used "average" values to aggregate numeric columns and "most frequent value" to aggregate categorical.

Here are my steps:

1. Aggregate the Heart Attack Dataset to ensure each Age-Gender pair has one unique record
2. Check the rows left after aggregation of Heart Attack Dataset
3. Aggregate the Obesity Dataset to ensure each Age-Gender pair has one unique record
4. Check the rows left after aggregation of the Obesity Dataset
5. Use the aggregate results to merge to find out the relationship between heart attack risk and obesity




### 5.1 Aggregate the Heart Attack Dataset
Summarize key metrics so that each Age-Gender pair has one unique record.

The left join (how="left") is used in the merge operation to ensure that all Age-Gender pairs from the numeric summary (heart_attack_summary) are preserved, even if there are missing values in the categorical summary (heart_attack_modes).

In [33]:
# Aggregating numeric features
heart_attack_summary = heart_attack.groupby(["Age", "Gender"]).agg({
    "BMI": "mean",  
    "Cholesterol_Level": "mean",  
    "Resting_BP": "mean",  
    "Heart_Rate": "mean", 
    "Max_Heart_Rate_Achieved": "mean"  
}).reset_index()

# Categorical columns to aggregate
categorical_cols = ["Smoking", "Alcohol_Consumption", "Diabetes", "Hypertension","Family_History", "Fasting_Blood_Sugar", "Exercise_Induced_Angina", 
                    "Physical_Activity_Level", "Stress_Level", "Chest_Pain_Type", 
                    "Thalassemia", "Fasting_Blood_Sugar", "ECG_Results", "Heart_Attack_Risk"]

# Aggregating categorical columns BEFORE merging with heart_attack_summary
heart_attack_modes = heart_attack.groupby(["Age", "Gender"])[categorical_cols].agg(lambda x: x.mode().iloc[0] if not x.mode().empty else None).reset_index()

# Merging numeric and categorical summaries
# Left join - the numeric summary of heart_attack_summary are more important, which should be preserved
heart_attack_summary = heart_attack_summary.merge(heart_attack_modes, on=["Age", "Gender"], how="left") 


# Display the aggregated dataset
heart_attack_summary.head(10)

Age  Gender        BMI  Cholesterol_Level  Resting_BP  Heart_Rate  \
0   18  Female  27.422077         222.865282  134.902077   95.121662   
1   18    Male  27.958696         226.126667  134.527536   93.573913   
2   19  Female  27.065015         222.713947  135.551929   93.919881   
3   19    Male  27.141158         223.723421  133.481579   95.807895   
4   20  Female  27.425770         222.471601  132.661631   95.966767   
5   20    Male  27.496226         227.285445  134.536388   93.846361   
6   21  Female  26.564234         225.058217  132.963788   94.083565   
7   21    Male  27.705086         225.191379  135.422414   96.563218   
8   22  Female  27.446677         225.296238  136.592476   95.344828   
9   22    Male  27.786276         224.057908  135.119898   94.913265   

   Max_Heart_Rate_Achieved  Smoking  Alcohol_Consumption  Diabetes  ...  \
0               149.857567    False                False     False  ...   
1               150.350725    False                False     False  ...   
2               149.709199    False                False     False  ...   
3               149.076316    False                False     False  ...   
4               151.456193    False                False     False  ...   
5               148.994609    False                False     False  ...   
6               150.417827    False                False     False  ...   
7               147.922414    False                False     False  ...   
8               148.884013    False                False     False  ...   
9               150.181122    False                False     False  ...   

   Family_History Fasting_Blood_Sugar  Exercise_Induced_Angina  \
0           False               False                    False   
1           False               False                    False   
2           False               False                    False   
3           False               False                    False   
4           False               False                    False   
5           False               False                    False   
6           False               False                    False   
7           False               False                    False   
8           False               False                    False   
9           False               False                    False   

   Physical_Activity_Level Stress_Level Chest_Pain_Type        Thalassemia  \
0                 Moderate     Moderate        Atypical             Normal   
1                 Moderate     Moderate         Typical             Normal   
2                 Moderate     Moderate        Atypical       Fixed defect   
3                 Moderate     Moderate         Typical  Reversible defect   
4                      Low     Moderate    Asymptomatic       Fixed defect   
5                      Low     Moderate    Asymptomatic             Normal   
6                      Low     Moderate        Atypical             Normal   
7                 Moderate     Moderate     Non-anginal  Reversible defect   
8                      Low     Moderate        Atypical  Reversible defect   
9                 Moderate     Moderate         Typical  Reversible defect   

  Fasting_Blood_Sugar                   ECG_Results Heart_Attack_Risk  
0               False                        Normal               Low  
1               False                        Normal               Low  
2               False  Left Ventricular Hypertrophy               Low  
3               False  Left Ventricular Hypertrophy               Low  
4               False                        Normal               Low  
5               False  Left Ventricular Hypertrophy               Low  
6               False                        Normal               Low  
7               False              ST-T abnormality               Low  
8               False                        Normal               Low  
9               False  Left Ventricular Hypertrophy               Low  

[10 r

### 5.2 Check the rows left after aggregation of the Heart Attack Dataset

Only 144 rows are left in the Heart Attack Dataset after the aggregate Age-Gender pair.

In [35]:
heart_attack_summary.shape[0]

144

### 5.3 Aggregate the Obesity Dataset
Similarly, aggregate relevant obesity data:

In [37]:
obesity_summary = obesity_clean.groupby(["Age", "Gender"]).agg({
    # Aggregating numerical columns with mean or median
    "Height": "mean",  # Average height
    "Weight": "mean",  # Average weight
    "FCVC": "mean",  # Average frequency of vegetable consumption
    "NCP": "mean",  # Average number of main meals per day
    "CH2O": "mean",  # Average water intake (daily)
    "FAF": "mean",  # Average frequency of physical activity
    "TUE": "mean",  # Average screen time (daily)
}).reset_index()

# Aggregating categorical columns with mode (most frequent value)
categorical_cols = ["Family_History", "FAVC", "CAEC", "Smoking", "SCC", "CALC", "MTRANS", "Obesity"]

for col in categorical_cols:
    obesity_summary[col] = obesity_clean.groupby(["Age", "Gender"])[col].agg(lambda x: x.mode().iloc[0] if not x.mode().empty else None).reset_index(drop=True)

# Display the aggregated dataset
obesity_summary.head(10)

Age  Gender    Height     Weight      FCVC       NCP      CH2O       FAF  \
0   14    Male  1.710000  72.000000  3.000000  3.000000  3.000000  2.000000   
1   15  Female  1.650000  86.000000  3.000000  3.000000  1.000000  3.000000   
2   16  Female  1.653906  59.489394  2.279552  2.080899  1.755732  1.160366   
3   16    Male  1.744161  53.104896  2.218404  2.365078  2.572868  1.578089   
4   17  Female  1.632780  57.195409  2.282397  2.732056  1.673945  0.941774   
5   17    Male  1.774628  67.982534  2.329973  3.259406  2.126902  1.599168   
6   18  Female  1.659497  79.886782  2.386031  2.751062  1.762544  1.032000   
7   18    Male  1.762358  76.113636  2.285235  2.819556  2.061293  1.154759   
8   19  Female  1.629894  66.336224  2.577224  2.629601  1.829438  1.221985   
9   19    Male  1.774745  78.160737  2.232072  2.990179  2.248056  1.557201   

        TUE  Family_History  FAVC       CAEC  Smoking    SCC       CALC  \
0  1.000000            True  True  Sometimes    False  False         no   
1  2.000000            True  True  Sometimes    False  False         no   
2  1.116546           False  True  Sometimes    False  False         no   
3  0.778753           False  True  Sometimes    False  False  Sometimes   
4  1.377729           False  True  Sometimes    False  False  Sometimes   
5  0.779923            True  True  Sometimes    False  False         no   
6  0.887597            True  True  Sometimes    False  False  Sometimes   
7  0.931332            True  True  Sometimes    False  False  Sometimes   
8  0.772383           False  True  Sometimes    False  False  Sometimes   
9  0.707325            True  True  Sometimes    False  False  Sometimes   

                  MTRANS              Obesity  
0                Walking        Normal_Weight  
1                Walking       Obesity_Type_I  
2  Public_Transportation   Overweight_Level_I  
3  Public_Transportation  Insufficient_Weight  
4  Public_Transportation        Normal_Weight  
5  Public_Transportation  Insufficient_Weight  
6  Public_Transportation     Obesity_Type_III  
7  Public_Transportation  Insufficient_Weight  
8  Public_Transportation  Insufficient_Weight  
9  Public_Transportation   Overweight_Level_I

### 5.4 Check the rows were left after aggregation of the Obesity Dataset

Only 68 rows are left in the Obesity Dataset after the aggregate Age-Gender pair.

In [39]:
obesity_summary.shape[0]

68

## 6. Merge datasets with common keys 

After ensuring "Age" and "Gender" are the same data types and removed duplicates, I merged two datasets on "Age" and "Gender" with "INNER JOIN".

#### - Why Use INNER JOIN?

__First, I only want patients present in both datasets to analyze the relationship between heart attack and obesity.__

An INNER JOIN keeps only the rows where there is a match in both datasets. 

- If someone appears in the heart attack dataset but not in the obesity dataset, I won't know their diet/exercise habits.
- If someone appears in the obesity dataset but not in the heart attack dataset, I won’t know their heart risk.

__Secondly, Keeping only complete records ensures better data quality. It avoids merging with missing data.__

- A LEFT JOIN would keep all heart attack records, even if no matching obesity data exists, leading to missing values in obesity-related columns.
- A RIGHT JOIN would keep all obesity records, but many wouldn’t have heart attack risk values.
- An INNER JOIN ensures every row has both heart risk and obesity data, reducing missing values.

In [42]:
# Merge the datasets on Age and Gender
merged_df = heart_attack_summary.merge(obesity_summary, on=["Age","Gender"], how="inner")  

In [43]:
# Display the first 20 rows of the merged dataset
merged_df.head(20)

Age  Gender        BMI  Cholesterol_Level  Resting_BP  Heart_Rate  \
0    18  Female  27.422077         222.865282  134.902077   95.121662   
1    18    Male  27.958696         226.126667  134.527536   93.573913   
2    19  Female  27.065015         222.713947  135.551929   93.919881   
3    19    Male  27.141158         223.723421  133.481579   95.807895   
4    20  Female  27.425770         222.471601  132.661631   95.966767   
5    20    Male  27.496226         227.285445  134.536388   93.846361   
6    21  Female  26.564234         225.058217  132.963788   94.083565   
7    21    Male  27.705086         225.191379  135.422414   96.563218   
8    22  Female  27.446677         225.296238  136.592476   95.344828   
9    22    Male  27.786276         224.057908  135.119898   94.913265   
10   23  Female  27.443571         224.955429  134.774286   94.374286   
11   23    Male  27.709435         225.216384  136.039548   94.401130   
12   24  Female  27.649354         221.990449  136.084270   92.011236   
13   24    Male  28.187468         222.883654  135.086538   95.535256   
14   25  Female  27.054550         224.904087  136.416894   95.411444   
15   25    Male  27.669006         227.481366  133.757764   94.198758   
16   26  Female  26.764314         229.072286  131.991429   94.994286   
17   26    Male  27.497003         220.992157  135.291317   93.803922   
18   27  Female  27.682832         223.144838  133.407080   94.734513   
19   27    Male  27.525228         227.409383  133.061662   94.265416   

    Max_Heart_Rate_Achieved  Smoking_x  Alcohol_Consumption  Diabetes  ...  \
0                149.857567      False                False     False  ...   
1                150.350725      False                False     False  ...   
2                149.709199      False                False     False  ...   
3                149.076316      False                False     False  ...   
4                151.456193      False                False     False  ...   
5                148.994609      False                False     False  ...   
6                150.417827      False                False     False  ...   
7                147.922414      False                False     False  ...   
8                148.884013      False                False     False  ...   
9                150.181122      False                False     False  ...   
10               149.917143      False                False     False  ...   
11               149.115819      False                False     False  ...   
12               149.213483      False                False     False  ...   
13               148.493590      False                False     False  ...   
14               151.422343      False                False     False  ...   
15               149.767081      False                False     False  ...   
16               151.591429      False                False     False  ...   
17               148.193277      False                False     False  ...   
18               149.510324      False                False     False  ...   
19               148.686327      False                False     False  ...   

         FAF       TUE  Family_History_y  FAVC       CAEC Smoking_y    SCC  \
0   1.032000  0.887597              True  True  Sometimes     False  False   
1   1.154759  0.931332              True  True  Sometimes     False  False   
2   1.221985  0.772383             False  True  Sometimes     False  False   
3   1.557201  0.707325              True  True  Sometimes     False  False   
4   1.164486  0.744071              True  True  Sometimes     False  False   
5   1.140084  0.770694              True  True  Sometimes     False  False   
6   1.216622  0.657769              True  True  Sometimes     False  False   
7   1.114029  0.577001              True  True  Sometimes     False  False   
8   0.933835  0.811006              True  True  Sometimes     False  False   
9   1.143854  0.791120              True  True  Som

As I observed before, the heart attack dataset has 144 rows and the Obesity dataset has 68 rows. 

__After joining, the merged dataset has 62 rows.__

__After merging, there are zero Na values.__

In [45]:
merged_df.describe(include = "all")

Age  Gender        BMI  Cholesterol_Level  Resting_BP  \
count   62.000000      62  62.000000          62.000000   62.000000   
unique        NaN       2        NaN                NaN         NaN   
top           NaN  Female        NaN                NaN         NaN   
freq          NaN      31        NaN                NaN         NaN   
mean    33.870968     NaN  27.489281         225.008328  134.356435   
std     10.529403     NaN   0.364873           2.471284    1.308569   
min     18.000000     NaN  26.564234         220.156906  131.091954   
25%     25.250000     NaN  27.279413         223.097429  133.351100   
50%     33.000000     NaN  27.539121         225.006823  134.645215   
75%     40.750000     NaN  27.732071         226.615902  135.257867   
max     61.000000     NaN  28.187468         231.002029  136.592476   

        Heart_Rate  Max_Heart_Rate_Achieved Smoking_x Alcohol_Consumption  \
count    62.000000                62.000000        62                  62   
unique         NaN                      NaN         1                   1   
top            NaN                      NaN     False               False   
freq           NaN                      NaN        62                  62   
mean     94.593195               149.614210       NaN                 NaN   
std       1.072339                 1.492698       NaN                 NaN   
min      92.011236               145.717143       NaN                 NaN   
25%      93.885233               148.700480       NaN                 NaN   
50%      94.445818               149.460329       NaN                 NaN   
75%      95.394790               150.385952       NaN                 NaN   
max      97.305476               153.550432       NaN                 NaN   

       Diabetes  ...        FAF        TUE Family_History_y  FAVC       CAEC  \
count        62  ...  62.000000  62.000000               62    62         62   
unique        1  ...        NaN        NaN                2     2          3   
top       False  ...        NaN        NaN             True  True  Sometimes   
freq         62  ...        NaN        NaN               58    58         60   
mean        NaN  ...   0.876660   0.413826              NaN   NaN        NaN   
std         NaN  ...   0.474544   0.303417              NaN   NaN        NaN   
min         NaN  ...   0.000000   0.000000              NaN   NaN        NaN   
25%         NaN  ...   0.576625   0.124614              NaN   NaN        NaN   
50%         NaN  ...   0.921112   0.401735              NaN   NaN        NaN   
75%         NaN  ...   1.152033   0.649705              NaN   NaN        NaN   
max         NaN  ...   2.000000   1.000000              NaN   NaN        NaN   

       Smoking_y    SCC       CALC      MTRANS         Obesity  
count         62     62         62          62              62  
unique         2      2          3           3               7  
top        False  False  Sometimes  Automobile  Obesity_Type_I  
freq          58     61         40          39              18  
mean         NaN    NaN        NaN         NaN             NaN  
std          NaN    NaN        NaN         NaN             NaN  
min          NaN    NaN        NaN         NaN             NaN  
25%          NaN    NaN        NaN         NaN             NaN  
50%          NaN    NaN        NaN         NaN             NaN  
75%          NaN    NaN        NaN         NaN             NaN  
max          NaN    NaN        NaN         NaN             NaN  

[11 rows x 36 columns]

In [46]:
merged_df.isna().sum().sum()

0

In [47]:
# Export the clean dataset
from datetime import date #importing the datetime class
import os
# Define the folder path
folder_path = "../clean_data/"

# Ensure the folder exists
os.makedirs(folder_path, exist_ok=True) 

file_name = "Sijian_Liang_Merged Data_"+ date.today().isoformat()+'.csv'

merged_df.to_csv(os.path.join(folder_path, file_name), index=False)

*End of the Assignment 5*

*For BABI 4005 DQM with Python*

*By: ITMG 4J Sijian Liang (Cindy)*
